File: src/commands/doctor.ts
import { readdirSync, readFileSync, existsSync } from 'fs';
import { join } from 'path';
import chalk from 'chalk';
import { parseSkillYaml, parseAgentYaml } from '../utils/yaml-loader.js';
import { lintSkill, type LintResult } from '../linters/rules.js';
import { checkDependencies, type DependencyIssue } from '../analyzers/dependency-checker.js';
import { detectLoopRisks, type LoopRisk } from '../analyzers/loop-detector.js';
import { checkSkillOrdering, type OrderingIssue } from '../analyzers/skill-ordering.js';
import type { SkillBehavior } from '../model/skill-behavior.js';

export interface DoctorReport {
  skills: SkillBehavior[];
  parseErrors: Array<{ file: string; error: string }>;
  lintIssues: Map<string, LintResult[]>;
  dependencyIssues: DependencyIssue[];
  loopRisks: Map<string, LoopRisk[]>;
  orderingIssues: OrderingIssue[];
  score: number;
}

export function runDoctor(skillsDir: string, agentsDir?: string): DoctorReport {
  const files = existsSync(skillsDir)
    ? readdirSync(skillsDir).filter((f) => f.endsWith('.skill.yaml'))
    : [];
  const skills: SkillBehavior[] = [];
  const parseErrors: DoctorReport['parseErrors'] = [];
  const lintIssues = new Map<string, LintResult[]>();
  const loopRisks = new Map<string, LoopRisk[]>();

  // Parse all skills
  for (const file of files) {
    const content = readFileSync(join(skillsDir, file), 'utf-8');
    const parsed = parseSkillYaml(content);
    if (parsed.success) {
      skills.push(parsed.data);
    } else {
      parseErrors.push({ file, error: parsed.error });
    }
  }

  // Lint each skill
  for (const skill of skills) {
    const issues = lintSkill(skill);
    if (issues.length > 0) lintIssues.set(skill.skill, issues);
  }

  // Check dependencies across all skills
  const dependencyIssues = checkDependencies(skills);

  // Check loop risks per skill
  for (const skill of skills) {
    const risks = detectLoopRisks(skill);
    if (risks.length > 0) loopRisks.set(skill.skill, risks);
  }

  // Check skill ordering in agents
  const orderingIssues: OrderingIssue[] = [];
  if (agentsDir && existsSync(agentsDir)) {
    const skillMap = new Map(skills.map((s) => [s.skill, s]));
    const agentFiles = readdirSync(agentsDir).filter((f) => f.endsWith('.agent.yaml'));
    for (const file of agentFiles) {
      const content = readFileSync(join(agentsDir, file), 'utf-8');
      const parsed = parseAgentYaml(content);
      if (parsed.success) {
        orderingIssues.push(...checkSkillOrdering(parsed.data, skillMap));
      }
    }
  }

  // Calculate health score
  const totalIssues =
    parseErrors.length +
    [...lintIssues.values()].flat().filter((i) => i.severity === 'error').length +
    dependencyIssues.length +
    [...loopRisks.values()].flat().filter((r) => r.severity === 'error').length +
    orderingIssues.length;

  const maxScore = Math.max(files.length * 10, 100);
  const score = Math.max(0, Math.round(100 - (totalIssues / maxScore) * 100));

  return { skills, parseErrors, lintIssues, dependencyIssues, loopRisks, orderingIssues, score };
}

export function printDoctorReport(report: DoctorReport): void {
  console.log(chalk.bold('\n=== Forgent Doctor Report ===\n'));
  console.log(`Skills found: ${report.skills.length}`);

  if (report.parseErrors.length > 0) {
    console.log(chalk.red(`\nParse Errors (${report.parseErrors.length}):`));
    for (const err of report.parseErrors) {
      console.log(`  ${chalk.red('x')} ${err.file}: ${err.error}`);
    }
  }

  if (report.lintIssues.size > 0) {
    console.log(chalk.yellow(`\nLint Issues:`));
    for (const [skill, issues] of report.lintIssues) {
      for (const issue of issues) {
        const icon = issue.severity === 'error' ? chalk.red('x') : chalk.yellow('!');
        console.log(`  ${icon} ${skill}: ${issue.message}`);
      }
    }
  }

  if (report.dependencyIssues.length > 0) {
    console.log(chalk.red(`\nDependency Issues (${report.dependencyIssues.length}):`));
    for (const issue of report.dependencyIssues) {
      console.log(`  ${chalk.red('x')} ${issue.skill}: ${issue.message}`);
    }
  }

  if (report.loopRisks.size > 0) {
    console.log(chalk.yellow(`\nLoop Risks:`));
    for (const [skill, risks] of report.loopRisks) {
      for (const risk of risks) {
        const icon = risk.severity === 'error' ? chalk.red('x') : chalk.yellow('!');
        console.log(`  ${icon} ${skill}: ${risk.message}`);
      }
    }
  }

  if (report.orderingIssues.length > 0) {
    console.log(chalk.yellow(`\nSkill Ordering Issues (${report.orderingIssues.length}):`));
    for (const issue of report.orderingIssues) {
      console.log(`  ${chalk.yellow('!')} ${issue.agent}: ${issue.message}`);
    }
  }

  const scoreColor = report.score >= 80 ? chalk.green : report.score >= 50 ? chalk.yellow : chalk.red;
  console.log(`\nHealth Score: ${scoreColor(`${report.score}/100`)}\n`);
}

File: internal/cmd/doctor.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/fatih/color"
	"github.com/mirandaguillaume/forgent/internal/analyzer"
	"github.com/mirandaguillaume/forgent/internal/linter"
	yamlloader "github.com/mirandaguillaume/forgent/internal/yaml"
	"github.com/mirandaguillaume/forgent/pkg/model"
	"github.com/spf13/cobra"
)

// DoctorReport holds the results of a full diagnostic run.
type DoctorReport struct {
	Skills           []model.SkillBehavior
	ParseErrors      []ParseError
	LintIssues       map[string][]linter.LintResult
	DependencyIssues []analyzer.DependencyIssue
	LoopRisks        map[string][]analyzer.LoopRisk
	OrderingIssues   []analyzer.OrderingIssue
	Score            int
}

// ParseError represents a skill file that failed to parse.
type ParseError struct {
	File  string
	Error string
}

// RunDoctor performs a full diagnostic on all skills and agents.
func RunDoctor(skillsDir string, agentsDir string) DoctorReport {
	report := DoctorReport{
		LintIssues: make(map[string][]linter.LintResult),
		LoopRisks:  make(map[string][]analyzer.LoopRisk),
	}

	// Parse all skills
	entries, _ := os.ReadDir(skillsDir)

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".skill.yaml") {
			continue
		}
		content, err := os.ReadFile(filepath.Join(skillsDir, entry.Name()))
		if err != nil {
			continue
		}
		skill, err := yamlloader.ParseSkillYAML(string(content))
		if err != nil {
			report.ParseErrors = append(report.ParseErrors, ParseError{
				File:  entry.Name(),
				Error: err.Error(),
			})
			continue
		}
		report.Skills = append(report.Skills, skill)
	}

	// Lint each skill
	for _, skill := range report.Skills {
		issues := linter.LintSkill(skill)
		if len(issues) > 0 {
			report.LintIssues[skill.Skill] = issues
		}
	}

	// Check dependencies across all skills
	report.DependencyIssues = analyzer.CheckDependencies(report.Skills)

	// Detect loop risks per skill
	checker := &analyzer.DefaultGuardrailChecker{}
	for _, skill := range report.Skills {
		risks := analyzer.DetectLoopRisks(skill, checker)
		if len(risks) > 0 {
			report.LoopRisks[skill.Skill] = risks
		}
	}

	// Check skill ordering in agents
	if agentsDir != "" {
		agentEntries, err := os.ReadDir(agentsDir)
		if err == nil {
			skillMap := make(map[string]model.SkillBehavior)
			for _, s := range report.Skills {
				skillMap[s.Skill] = s
			}
			for _, entry := range agentEntries {
				if !strings.HasSuffix(entry.Name(), ".agent.yaml") {
					continue
				}
				content, err := os.ReadFile(filepath.Join(agentsDir, entry.Name()))
				if err != nil {
					continue
				}
				agent, err := yamlloader.ParseAgentYAML(string(content))
				if err != nil {
					continue
				}
				report.OrderingIssues = append(report.OrderingIssues,
					analyzer.CheckSkillOrdering(agent, skillMap)...)
			}
		}
	}

	// Calculate health score
	totalIssues := len(report.ParseErrors) + len(report.DependencyIssues) + len(report.OrderingIssues)
	for _, issues := range report.LintIssues {
		for _, i := range issues {
			if i.Severity == linter.SeverityError {
				totalIssues++
			}
		}
	}
	for _, risks := range report.LoopRisks {
		for _, r := range risks {
			if r.Severity == "error" {
				totalIssues++
			}
		}
	}
	fileCount := len(report.Skills) + len(report.ParseErrors)
	maxScore := fileCount * 10
	if maxScore < 100 {
		maxScore = 100
	}
	report.Score = max(0, 100-(totalIssues*100/maxScore))

	return report
}

// PrintDoctorReport prints the doctor report to stdout with colored output.
func PrintDoctorReport(report DoctorReport) {
	bold := color.New(color.Bold)
	fmt.Println()
	bold.Println("=== Forgent Doctor Report ===")
	fmt.Println()
	fmt.Printf("Skills found: %d\n", len(report.Skills))

	if len(report.ParseErrors) > 0 {
		fmt.Println(color.RedString("\nParse Errors (%d):", len(report.ParseErrors)))
		for _, err := range report.ParseErrors {
			fmt.Printf("  %s %s: %s\n", color.RedString("x"), err.File, err.Error)
		}
	}

	if len(report.LintIssues) > 0 {
		fmt.Println(color.YellowString("\nLint Issues:"))
		for skill, issues := range report.LintIssues {
			for _, issue := range issues {
				icon := color.YellowString("!")
				if issue.Severity == linter.SeverityError {
					icon = color.RedString("x")
				}
				fmt.Printf("  %s %s: %s\n", icon, skill, issue.Message)
			}
		}
	}

	if len(report.DependencyIssues) > 0 {
		fmt.Println(color.RedString("\nDependency Issues (%d):", len(report.DependencyIssues)))
		for _, issue := range report.DependencyIssues {
			fmt.Printf("  %s %s: %s\n", color.RedString("x"), issue.Skill, issue.Message)
		}
	}

	if len(report.LoopRisks) > 0 {
		fmt.Println(color.YellowString("\nLoop Risks:"))
		for skill, risks := range report.LoopRisks {
			for _, risk := range risks {
				icon := color.YellowString("!")
				if risk.Severity == "error" {
					icon = color.RedString("x")
				}
				fmt.Printf("  %s %s: %s\n", icon, skill, risk.Message)
			}
		}
	}

	if len(report.OrderingIssues) > 0 {
		fmt.Println(color.YellowString("\nSkill Ordering Issues (%d):", len(report.OrderingIssues)))
		for _, issue := range report.OrderingIssues {
			fmt.Printf("  %s %s: %s\n", color.YellowString("!"), issue.Agent, issue.Message)
		}
	}

	scoreColor := color.GreenString
	if report.Score < 80 {
		scoreColor = color.YellowString
	}
	if report.Score < 50 {
		scoreColor = color.RedString
	}
	fmt.Printf("\nHealth Score: %s\n\n", scoreColor("%d/100", report.Score))
}

func init() {
	var agentsDir string
	doctorCmd := &cobra.Command{
		Use:   "doctor [path]",
		Short: "Run full diagnostic on all skills and agents",
		Args:  cobra.MaximumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			path := "skills"
			if len(args) > 0 {
				path = args[0]
			}
			report := RunDoctor(path, agentsDir)
			PrintDoctorReport(report)
			if report.Score < 50 {
				os.Exit(1)
			}
		},
	}
	doctorCmd.Flags().StringVarP(&agentsDir, "agents", "a", "agents", "agents directory")
	rootCmd.AddCommand(doctorCmd)
}
